File: python/semantic_kernel/core_plugins/crew_ai/crew_ai_enterprise_client.py
# Copyright (c) Microsoft. All rights reserved.

from typing import Any

import aiohttp

from semantic_kernel.core_plugins.crew_ai.crew_ai_models import (
    CrewAIKickoffResponse,
    CrewAIRequiredInputs,
    CrewAIStatusResponse,
)
from semantic_kernel.utils.telemetry.user_agent import SEMANTIC_KERNEL_USER_AGENT


class CrewAIEnterpriseClient:
    """Client to interact with the Crew AI Enterprise API."""

    def __init__(
        self,
        endpoint: str,
        auth_token: str,
        session: aiohttp.ClientSession | None = None,
    ):
        """Initializes a new instance of the CrewAIEnterpriseClient class.

        Args:
            endpoint (str): The API endpoint.
            auth_token (str): The authentication token.
            session (aiohttp.ClientSession | None, optional): The HTTP client session. Defaults to None.
        """
        self.endpoint = endpoint
        self.auth_token = auth_token
        self.session = session if session is None else aiohttp.ClientSession()
        self.request_header = {
            "Authorization": f"Bearer {auth_token}",
            "Content-Type": "application/json",
            "user_agent": SEMANTIC_KERNEL_USER_AGENT,
        }

    async def get_inputs(self) -> CrewAIRequiredInputs:
        """Get the required inputs for Crew AI.

        Returns:
            CrewAIRequiredInputs: The required inputs for Crew AI.
        """
        async with (
            self.session.get(f"{self.endpoint}/inputs", headers=self.request_header) as response,  # type: ignore
        ):
            response.raise_for_status()
            return CrewAIRequiredInputs.model_validate_json(await response.text())

    async def kickoff(
        self,
        inputs: dict[str, Any] | None = None,
        task_webhook_url: str | None = None,
        step_webhook_url: str | None = None,
        crew_webhook_url: str | None = None,
    ) -> CrewAIKickoffResponse:
        """Kickoff a new Crew AI task.

        Args:
            inputs (Optional[dict[str, Any]], optional): The inputs for the task. Defaults to None.
            task_webhook_url (Optional[str], optional): The webhook URL for task updates. Defaults to None.
            step_webhook_url (Optional[str], optional): The webhook URL for step updates. Defaults to None.
            crew_webhook_url (Optional[str], optional): The webhook URL for crew updates. Defaults to None.

        Returns:
            CrewAIKickoffResponse: The response from the kickoff request.
        """
        content = {
            "inputs": inputs,
            "taskWebhookUrl": task_webhook_url,
            "stepWebhookUrl": step_webhook_url,
            "crewWebhookUrl": crew_webhook_url,
        }
        async with (
            self.session.post(f"{self.endpoint}/kickoff", json=content, headers=self.request_header) as response,  # type: ignore
        ):
            response.raise_for_status()
            body = await response.text()
            return CrewAIKickoffResponse.model_validate_json(body)

    async def get_status(self, task_id: str) -> CrewAIStatusResponse:
        """Get the status of a Crew AI task.

        Args:
            task_id (str): The ID of the task.

        Returns:
            CrewAIStatusResponse: The status response of the task.
        """
        async with (
            self.session.get(f"{self.endpoint}/status/{task_id}", headers=self.request_header) as response,  # type: ignore
        ):
            response.raise_for_status()
            body = await response.text()
            return CrewAIStatusResponse.model_validate_json(body)

    async def __aenter__(self):
        """Enter the session."""
        await self.session.__aenter__()  # type: ignore
        return self

    async def __aexit__(self, *args, **kwargs):
        """Close the session."""
        await self.session.close()  # type: ignore

File: dotnet/src/Plugins/Plugins.AI/CrewAI/Client/CrewAIEnterpriseClient.cs
// Copyright (c) Microsoft. All rights reserved.

using System;
using System.Net.Http;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.SemanticKernel.Http;

namespace Microsoft.SemanticKernel.Plugins.AI.CrewAI;

/// <summary>
/// Internal interface used for mocking and testing.
/// </summary>
internal interface ICrewAIEnterpriseClient
{
    Task<CrewAIRequiredInputs> GetInputsAsync(CancellationToken cancellationToken = default);
    Task<CrewAIKickoffResponse> KickoffAsync(
        object? inputs,
        string? taskWebhookUrl = null,
        string? stepWebhookUrl = null,
        string? crewWebhookUrl = null,
        CancellationToken cancellationToken = default);
    Task<CrewAIStatusResponse> GetStatusAsync(string taskId, CancellationToken cancellationToken = default);
}

/// <summary>
/// A client for interacting with the CrewAI Enterprise API.
/// </summary>
internal class CrewAIEnterpriseClient : ICrewAIEnterpriseClient
{
    private readonly Uri _endpoint;
    private readonly Func<Task<string>> _authTokenProvider;
    private readonly IHttpClientFactory? _httpClientFactory;

    public CrewAIEnterpriseClient(Uri endpoint, Func<Task<string>> authTokenProvider, IHttpClientFactory? clientFactory = null)
    {
        Verify.NotNull(endpoint, nameof(endpoint));
        Verify.NotNull(authTokenProvider, nameof(authTokenProvider));

        this._endpoint = endpoint;
        this._authTokenProvider = authTokenProvider;
        this._httpClientFactory = clientFactory;
    }

    /// <summary>
    /// Get the inputs required for the Crew to kickoff.
    /// </summary>
    /// <param name="cancellationToken">A <see cref="CancellationToken"/></param>
    /// <returns>Aninstance of <see cref="CrewAIRequiredInputs"/> describing the required inputs.</returns>
    /// <exception cref="KernelException"></exception>
    public async Task<CrewAIRequiredInputs> GetInputsAsync(CancellationToken cancellationToken = default)
    {
        try
        {
            using var client = await this.CreateHttpClientAsync().ConfigureAwait(false);
            using var requestMessage = HttpRequest.CreateGetRequest("/inputs");
            using var response = await client.SendWithSuccessCheckAsync(requestMessage, cancellationToken)
                .ConfigureAwait(false);

            var body = await response.Content.ReadAsStringWithExceptionMappingAsync(cancellationToken)
                .ConfigureAwait(false);

            var requirements = JsonSerializer.Deserialize<CrewAIRequiredInputs>(body);

            return requirements ?? throw new KernelException(message: $"Failed to deserialize requirements from CrewAI. Response: {body}");
        }
        catch (Exception ex) when (ex is not KernelException)
        {
            throw new KernelException(message: "Failed to get required inputs for CrewAI Crew.", innerException: ex);
        }
    }

    /// <summary>
    /// Kickoff the Crew.
    /// </summary>
    /// <param name="inputs">An object containing key value pairs matching the required inputs of the Crew.</param>
    /// <param name="taskWebhookUrl">The task level webhook Uri.</param>
    /// <param name="stepWebhookUrl">The step level webhook Uri.</param>
    /// <param name="crewWebhookUrl">The crew level webhook Uri.</param>
    /// <param name="cancellationToken">A <see cref="CancellationToken"/></param>
    /// <returns>A string containing the Id of the started Crew Task.</returns>
    public async Task<CrewAIKickoffResponse> KickoffAsync(
        object? inputs,
        string? taskWebhookUrl = null,
        string? stepWebhookUrl = null,
        string? crewWebhookUrl = null,
        CancellationToken cancellationToken = default)
    {
        try
        {
            var content = new
            {
                inputs,
                taskWebhookUrl,
                stepWebhookUrl,
                crewWebhookUrl
            };

            using var client = await this.CreateHttpClientAsync().ConfigureAwait(false);
            using var requestMessage = HttpRequest.CreatePostRequest("/kickoff", content);
            using var response = await client.SendWithSuccessCheckAsync(requestMessage, cancellationToken)
                .ConfigureAwait(false);

            var body = await response.Content.ReadAsStringWithExceptionMappingAsync(cancellationToken)
                .ConfigureAwait(false);

            var kickoffResponse = JsonSerializer.Deserialize<CrewAIKickoffResponse>(body);
            return kickoffResponse ?? throw new KernelException(message: $"Failed to deserialize kickoff response from CrewAI. Response: {body}");
        }
        catch (Exception ex) when (ex is not KernelException)
        {
            throw new KernelException(message: "Failed to kickoff CrewAI Crew.", innerException: ex);
        }
    }

    /// <summary>
    /// Get the status of the Crew Task.
    /// </summary>
    /// <param name="taskId">The Id of the task.</param>
    /// <param name="cancellationToken">A <see cref="CancellationToken"/></param>
    /// <returns>A string containing the status or final result of the Crew task.</returns>
    /// <exception cref="KernelException"></exception>
    public async Task<CrewAIStatusResponse> GetStatusAsync(string taskId, CancellationToken cancellationToken = default)
    {
        try
        {
            using var client = await this.CreateHttpClientAsync().ConfigureAwait(false);
            using var requestMessage = HttpRequest.CreateGetRequest($"/status/{taskId}");
            using var response = await client.SendWithSuccessCheckAsync(requestMessage, cancellationToken)
                .ConfigureAwait(false);

            var body = await response.Content.ReadAsStringWithExceptionMappingAsync(cancellationToken)
                .ConfigureAwait(false);

            var statusResponse = JsonSerializer.Deserialize<CrewAIStatusResponse>(body);

            return statusResponse ?? throw new KernelException(message: $"Failed to deserialize status response from CrewAI. Response: {body}");
        }
        catch (Exception ex) when (ex is not KernelException)
        {
            throw new KernelException(message: "Failed to status of CrewAI Crew.", innerException: ex);
        }
    }

    #region Private Methods

    private async Task<HttpClient> CreateHttpClientAsync()
    {
        var authToken = await this._authTokenProvider().ConfigureAwait(false);

        if (string.IsNullOrWhiteSpace(authToken))
        {
            throw new KernelException(message: "Failed to get auth token for CrewAI.");
        }

        var client = this._httpClientFactory?.CreateClient() ?? new();
        client.DefaultRequestHeaders.Add("Authorization", $"Bearer {authToken}");
        client.BaseAddress = this._endpoint;
        return client;
    }

    #endregion
}
